Make the compilation cache key and the CAS outputs data output path independent

akyrtzi opened the pull request:
Instead of storing CAS outputs by the specific path that was used at the time of invocation, associate them with an "output kind name" so that they are re-usable even with different output paths.

The practical benefit of this is that we can get cache hits when source files are part of multiple products (e.g. when source files are included in multiple Xcode test targets) or when PCH files end up with the same caching key and output but the build system is not in a position identify this.

akyrtzi committed:
Make the compilation cache key and the CAS outputs data output path independent

Instead of storing CAS outputs by the specific path that was used at the time of invocation, associate them with an "output kind name" so that they are re-usable even with different output paths.

The practical benefit of this is that we can get cache hits when source files are part of multiple products (e.g. when source files are included in multiple Xcode test targets) or when PCH files end up with the same caching key and output but the build system is not in a position identify this.

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan reviewed (approved):
LGTM. 

I wonder if `MergeChildRecords` works with this cached serialized diagnostics. I think it should but it would be good to have a testcase for that. Basically you need a test that driver also emits a warning and see if the final serialized diag contains both driver warning and warning in CU.

cachemeifyoucan commented:
> I wonder if `MergeChildRecords` works with this cached serialized diagnostics. I think it should but it would be good to have a testcase for that. Basically you need a test that driver also emits a warning and see if the final serialized diag contains both driver warning and warning in CU.

Actually, I don't think this commit change this behavior at all. Might be better add the test in https://github.com/apple/llvm-project/pull/5042 instead.

LGTM

akyrtzi commented:
> I wonder if MergeChildRecords works with this cached serialized diagnostics. I think it should but it would be good to have a testcase for that.

This doesn't seem to work, the diagnostics from the actual compilation don't end up in the serialized diagnostics file. I'll look into fixing it as part of https://github.com/apple/llvm-project/pull/5042.

benlangmuir commented on clang/tools/driver/cc1_main.cpp:
```diff
@@ -235,6 +235,25 @@ namespace {
+  /// compilation cache key from the specific output paths.
+  enum class OutputKind {
+    MainOutput,
+    SerializedDiagnostics,
```
What about dependency file? When we go through the "full dependencies" API the dependency file is written by the compiler not the scanner.

akyrtzi replied:
> When we go through the "full dependencies" API the dependency file is written by the compiler not the scanner.

Are there tests that exercise this? I thought all the caching tests we have are using the driver's caching mechanism.

benlangmuir replied:
clang/test/CAS/fcache-compile-job-dependency-file.c uses cc1 commands, so it should exercise this without going through the driver/scanner.

akyrtzi replied:
Handling dependency files as well in updated PR, using `test/CAS/fcache-compile-job-dependency-file.c` for testing.

benlangmuir commented on clang/lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -56,18 +56,26 @@ clang::createCompileJobCacheKey(CASDB &CAS, ArrayRef<const char *> CC1Args,
+                                const CompilerInvocation &OriginalInvocation) {
+  CompilerInvocation InvocationForCacheKey(OriginalInvocation);
+  FrontendOptions &FrontendOpts = InvocationForCacheKey.getFrontendOpts();
+  DiagnosticOptions &DiagOpts = InvocationForCacheKey.getDiagnosticOpts();
```
It occurred to me that this creates a potential foot-gun when using the other overload of `createCompileJobCacheKey` that takes arguments directly, since only one of them does canonicalization.  Maybe we should get rid of the other one, or at least add a comment about the difference?

cachemeifyoucan replied:
The offline feedback I provide is that it would be good just to have CompilerInvocation to create compilation cache key, which just hashing over all the options and each options knows what are the outputs and exclude those from hashing. Not sure how complicated that is.

akyrtzi replied:
I removed the plain arguments variant with a [separate commit](https://github.com/apple/llvm-project/pull/5036/commits/3038da5863ffa80fdc1661929922c21c8c1f1b0e).

The only caller was the dependency scanning daemon, using it to do a check of whether the key is cached or not and track the cache hits number in a global variable. This particular use is debatable whether it provides enough value IMO.

akyrtzi commented:
> > I wonder if MergeChildRecords works with this cached serialized diagnostics. I think it should but it would be good to have a testcase for that.
> 
> This doesn't seem to work, the diagnostics from the actual compilation don't end up in the serialized diagnostics file. I'll look into fixing it as part of #5042.

I've opened a separate PR here: https://github.com/apple/llvm-project/pull/5078

benlangmuir commented on clang/lib/Frontend/CompileJobCacheKey.cpp:
```diff
@@ -56,18 +56,26 @@ clang::createCompileJobCacheKey(CASDB &CAS, ArrayRef<const char *> CC1Args,
+  DiagnosticOptions &DiagOpts = InvocationForCacheKey.getDiagnosticOpts();
+  // Keep the key independent of the paths of these outputs.
+  FrontendOpts.OutputFile.clear();
+  DiagOpts.DiagnosticSerializationFile.clear();
```
I think we need something more like
```c++
if (!DiagOpts.DiagnosticSerializationFile.empty())
  DiagOpts.DiagnosticSerializationFile = "-";
```

We can change the path of the output, but we cannot reuse a cached result that had no serialized diagnostics in the first place.  Alternatively, we could always generate serialized diagnostics when caching.

benlangmuir replied:
Note: this is similar to why `ModuleDepCollector` has to store flags for `HadSerializedDiagnostics` and `HadDependencyFile` and include those in the module context hash.

akyrtzi replied:
Good catch, and I like the idea of "always generate serialized diagnostics when caching".
Later on we can improve how we handle diagnostics per [this FIXME](https://github.com/apple/llvm-project/pull/5078/commits/58f2dd0701e19fa9a28480b2531a66154420763c#diff-c6f8f185d25be2e8d40d9c42ffcca1caac9b2f8403da61bd7957b6957763cb86R433).

akyrtzi replied:
See updated PR.

akyrtzi committed:
[CompileJobCacheKey] Remove the `createCompileJobCacheKey()` function that accepts an array of arguments

Only expose the variant of `createCompileJobCacheKey()` that accepts a `CompilerInvocation` and does canonicalization
for the key calculation, to make sure there is no mistake with a caller passing the wrong set of arguments for the other variant.

The only caller of the arguments variant was the dependency scanning daemon, using it to keep track of a global statistics
variable `NumCASCacheHit` (by doing an otherwise unnecessary cache key lookup).
It is not clear that the daemon keeping track of that variable is important, we could design a better scheme for recording the cache hit rate.

akyrtzi commented:
@swift-ci Please test

benlangmuir commented on clang/tools/driver/cc1_main.cpp:
```diff
@@ -469,6 +537,14 @@ Optional<int> CompileJobCache::tryReplayCachedResult(CompilerInstance &Clang) {
         Diags.takeClient(), std::move(SerializedConsumer)));
+  } else {
+    // We always generate the serialized diagnostics so the key is independent
+    // of the presence of '--serialize-diagnostics'.
```
Do we need to always generate the .d file when caching for the same reason?

akyrtzi replied:
It's not clear to me how valuable that is, the implicit clang driver caching method is generating the .d file during dep-scanning (not cached), and when using the libclang API, generation of .d files should become unnecessary at some point.

benlangmuir replied:
I'm fine with not optimizing for this case, but we should make it part of the cache key then, not just silently don't generate the .d file when replaying.  I suggest if there is a path, map it to "-", but if there is no path leave it empty.

cachemeifyoucan replied:
`.d` file is generated by dependency scanner so it should never be emitted from main compilation in cached build.

benlangmuir replied:
This is true for in the driver+daemon, but not when using libclang APIs.

akyrtzi replied:
> I suggest if there is a path, map it to "-", but if there is no path leave it empty.

Good point, I've updated the PR with such a change.

akyrtzi commented:
@swift-ci Please test
